src/Storages/ExternalTable/ClickHouse/ClickHouse.cpp

```diff
@@ -0,0 +1,125 @@
+    if (!port)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Invalid port in ClickHouse address");
+
+    connection_params.host = host;
+    connection_params.port = port;
+    connection_params.user = settings->user.value;
+    connection_params.password = settings->password.value;
+    connection_params.default_database = settings->database.value;
+    connection_params.security = settings->secure.value ? Protocol::Secure::Enable : Protocol::Secure::Disable;
+    /// TODO read from settings
+    connection_params.timeouts = {
+        /*connection_timeout_=*/ 1 * 60 * 1'000'000,
+        /*send_timeout_=*/ 1 * 60 * 1'000'000,
+        /*receive_timeout_=*/ 1 * 60 * 1'000'000,
+        /*tcp_keep_alive_timeout_=*/ 5 * 60 * 1'000'000
+    };
+}
+
+void ClickHouse::startup()
+{
+    LOG_INFO(logger, "startup");
+}
+
+Pipe ClickHouse::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo &  /*query_info*/,
+    ContextPtr context,
+    QueryProcessingStage::Enum processed_stage,
+    size_t  /*max_block_size*/,
+    size_t /*num_streams*/)
+{
+    auto header = storage_snapshot->getSampleBlockForColumns(column_names);
+    auto client = std::make_unique<ClickHouseClient>(connection_params, logger);
+    auto source = std::make_shared<ClickHouseSource>(database, table, std::move(header), std::move(client), processed_stage, context, logger);
+    return Pipe(std::move(source));
+}
+
+SinkToStoragePtr ClickHouse::write(const ASTPtr &  /*query*/, const StorageMetadataPtr & metadata_snapshot, ContextPtr  context)
+{
+    return std::make_shared<ClickHouseSink>(database, table, metadata_snapshot->getSampleBlock(), connection_params, context, logger);
+}
+
+ColumnsDescription ClickHouse::getTableStructure()
+{
+    ColumnsDescription ret {};
+
+    ClickHouseClient client {connection_params, logger};
+    auto query = "DESCRIBE TABLE " + (database.empty() ? "" : backQuoteIfNeed(database) + ".") + backQuoteIfNeed(table);
+    /// This has to fail quickly otherwise it will block Proton from starting.
+    client.executeQuery(query, /*query_id=*/"", /*fail_quick=*/true);
+    LOG_INFO(logger, "Receiving table schema");
+    while (true)
+    {
+        const auto & block = client.pollData();
+        if (!block)
+            break;
+
+        auto rows = block->rows();
+        if (!rows)
+            continue;
+
+        const auto & cols = block.value().getColumns();
+        const auto & factory = DataTypeFactory::instance();
+        for (size_t i = 0; i < rows; ++i)
+        {
+            ColumnDescription col_desc {};
+            {
+                const auto & col = block->getByName("name");
+                col_desc.name = col.column->getDataAt(i).toString();
+            }
+            {
+                const auto & col = block->getByName("type");
+                col_desc.type = factory.get(col.column->getDataAt(i).toString(), true);
```

Comment:
We can explicitly comment for the `true` parameter like `/*is_clickhouse_compatible=*/ true`